dwave/optimization/src/graph.cpp

```diff
@@ -298,6 +298,71 @@ void Graph::propose(State& state, std::vector<const Node*> sources,
     }
 }
 
+ssize_t Graph::remove_unused_nodes(bool ignore_listeners) {
+    // Establish a topological ordering. We'll use the fact that the node list
+    // is ordered, but we're going to mess with the topological indices!
+    topological_sort();
+
+    // We'll use this later to calculate how many nodes we removed
+    const ssize_t num_nodes_before = this->num_nodes();
+
+    // We generally want to avoid touching decisions
+    const ssize_t num_decisions = this->num_decisions();
+
+    // Mark the nodes that we plan to keep regardless of their number of successors
+    // For performance we store a signalling value in the topological_index of the
+    // nodes. We need to take some care to not override the topological index of
+    // any decisions.
+    constexpr ssize_t keep = -2;
```

Comment:
This is set to `-2` because `-1` already has a meaning of "unsorted", correct? But at this point you've topologically sorted so it should be guaranteed all nodes have an index >= 0.

Reply:
That's right. I just need a negative number. I chose `-2` because `-1` already has meaning as unsorted. Should never matter but :shrug: